dev/com.ibm.ws.security.common.jsonwebkey/src/com/ibm/ws/security/common/jwk/impl/JWKProvider.java

```diff
@@ -0,0 +1,216 @@
+
+    protected String alg = null;
+    protected String use = null;
+    protected int size = DEFAULT_KEY_SIZE;
+    protected Timer timer;
+    protected long rotationTimeInMilliseconds = DEFAULT_ROTATION_TIME;
+
+    protected PublicKey publicKey = null;
+    protected PrivateKey privateKey = null;
+
+    protected JWKProvider() {
+        this(DEFAULT_KEY_SIZE, RS256, DEFAULT_ROTATION_TIME);
+    }
+
+    public JWKProvider(int keySize, String alg, long rotationTimeMs) {
+        if (keySize < 0) {
+            if (tc.isDebugEnabled()) {
+                Tr.debug(tc, "Specified key size " + keySize + " < 0. Setting key size to the default (" + DEFAULT_KEY_SIZE + ") instead");
+            }
+            keySize = DEFAULT_KEY_SIZE;
+        }
+        this.size = keySize;
+        this.alg = alg;
+        if (rotationTimeMs <= 0) {
+            if (tc.isDebugEnabled()) {
+                Tr.debug(tc, "Specified rotation time " + rotationTimeMs + " <= 0. Setting rotation time to the default (" + DEFAULT_ROTATION_TIME + " ms) instead");
+            }
+            rotationTimeMs = DEFAULT_ROTATION_TIME;
+        }
+        this.rotationTimeInMilliseconds = rotationTimeMs;
+
+        scheduleRotationTask();
+    }
+
+    public JWKProvider(int keySize, String alg, long rotationTimeMs, PublicKey publicKey, PrivateKey privateKey) {
+        if (keySize < 0) {
+            if (tc.isDebugEnabled()) {
+                Tr.debug(tc, "Specified key size " + keySize + " < 0. Setting key size to the default (" + DEFAULT_KEY_SIZE + ") instead");
+            }
+            keySize = DEFAULT_KEY_SIZE;
+        }
+        this.size = keySize;
+        this.alg = alg;
+        if (rotationTimeMs <= 0) {
+            if (tc.isDebugEnabled()) {
+                Tr.debug(tc, "Specified rotation time " + rotationTimeMs + " <= 0. Setting rotation time to the default (" + DEFAULT_ROTATION_TIME + " ms) instead");
+            }
+            rotationTimeMs = DEFAULT_ROTATION_TIME;
+        }
+        this.rotationTimeInMilliseconds = rotationTimeMs;
+        //this.jwkEnabled = jwkEnabled;
+        this.publicKey = publicKey;
+        this.privateKey = privateKey;
+    }
+
+    public JSONWebKey getJWK() {
+        JWK jwk = null;
+        while (jwks.size() < JWKS_TO_GENERATE) {
+            generateJWKs();
+        }
+        jwk = jwks.get(0);
+        return jwk;
+    }
+
+    protected void generateJWKs() {
+        JWK jwk = null;
+        while (jwks.size() < JWKS_TO_GENERATE) {
+            jwk = generateJWK(alg, size);
+            jwks.add(jwk);
+        }
+    }
+
+    protected JWK generateJWK(String alg, int size) {
+        JWK jwk = null;
+        if (RS256.equals(alg)) {
+            if (publicKey != null && privateKey != null) {
+                jwk = Jose4jRsaJWK.getInstance(alg, use, publicKey, privateKey);
+                jwk.generateKey();
+            } else {
+                jwk = generateRsaJWK(alg, size);
+            }
+        }
```

Comment:
might want to throw an UnsupportedOperationException if alg!=256 rather than just failing quietly?  (I've decided those are easier than adding messages)

Reply:
Thanks Bruce. This code will be changed to support other algorithms. I will update it when we add the new code.